File: python-backend/monitoring/metrics.py
"""
Metrics collection system for Python backend
"""

import time
import json
from datetime import datetime
from typing import Dict, Any, List, Optional
from dataclasses import dataclass, asdict
from collections import defaultdict, deque
from .logger import logger

@dataclass
class MetricData:
    name: str
    value: float
    unit: str
    timestamp: str
    tags: Dict[str, str]

@dataclass
class AgentMetrics:
    agent_id: str
    agent_type: str
    tenant_id: str
    workflow_id: Optional[str]
    operation: str
    duration: float
    status: str  # 'success', 'error', 'timeout'
    input_tokens: Optional[int] = None
    output_tokens: Optional[int] = None
    rag_queries: Optional[int] = None

@dataclass
class SystemMetrics:
    cpu_usage: Optional[float] = None
    memory_usage: Optional[float] = None
    active_agents: int = 0
    active_tenants: int = 0
    error_rate: float = 0.0
    response_time: float = 0.0

class MetricsCollector:
    def __init__(self, max_metrics: int = 10000):
        self.metrics: deque = deque(maxlen=max_metrics)
        self.max_metrics = max_metrics

    def _create_metric(
        self,
        name: str,
        value: float,
        unit: str,
        tags: Optional[Dict[str, str]] = None
    ) -> MetricData:
        return MetricData(
            name=name,
            value=value,
            unit=unit,
            timestamp=datetime.utcnow().isoformat(),
            tags=tags or {}
        )

    def _add_metric(self, metric: MetricData) -> None:
        self.metrics.append(metric)
        
        # Log metric for external collection
        logger.info("Metric collected", {
            "component": "metrics",
            "metadata": asdict(metric)
        })

    def record_agent_operation(self, metrics: AgentMetrics) -> None:
        """Record metrics for agent operations"""
        tags = {
            "agent_id": metrics.agent_id,
            "agent_type": metrics.agent_type,
            "tenant_id": metrics.tenant_id,
            "operation": metrics.operation,
            "status": metrics.status
        }

        if metrics.workflow_id:
            tags["workflow_id"] = metrics.workflow_id

        # Record operation count
        self._add_metric(self._create_metric(
            "agent.operation",
            1,
            "count",
            tags
        ))

        # Record duration
        self._add_metric(self._create_metric(
            "agent.duration",
            metrics.duration,
            "milliseconds",
            tags
        ))

        # Record token usage if available
        if metrics.input_tokens is not None:
            self._add_metric(self._create_metric(
                "agent.tokens.input",
                metrics.input_tokens,
                "count",
                tags
            ))

        if metrics.output_tokens is not None:
            self._add_metric(self._create_metric(
                "agent.tokens.output",
                metrics.output_tokens,
                "count",
                tags
            ))

        if metrics.rag_queries is not None:
            self._add_metric(self._create_metric(
                "agent.rag.queries",
                metrics.rag_queries,
                "count",
                tags
            ))

    def record_system_metrics(self, metrics: SystemMetrics) -> None:
        """Record system-level metrics"""
        tags = {"component": "system"}

        if metrics.cpu_usage is not None:
            self._add_metric(self._create_metric(
                "system.cpu.usage",
                metrics.cpu_usage,
                "percent",
                tags
            ))

        if metrics.memory_usage is not None:
            self._add_metric(self._create_metric(
                "system.memory.usage",
                metrics.memory_usage,
                "bytes",
                tags
            ))

        self._add_metric(self._create_metric(
            "system.agents.active",
            metrics.active_agents,
            "count",
            tags
        ))

        self._add_metric(self._create_metric(
            "system.tenants.active",
            metrics.active_tenants,
            "count",
            tags
        ))

        self._add_metric(self._create_metric(
            "system.error.rate",
            metrics.error_rate,
            "percent",
            tags
        ))

        self._add_metric(self._create_metric(
            "system.response.time",
            metrics.response_time,
            "milliseconds",
            tags
        ))

    def record_custom_metric(
        self,
        name: str,
        value: float,
        unit: str,
        tags: Optional[Dict[str, str]] = None
    ) -> None:
        """Record a custom metric"""
        self._add_metric(self._create_metric(name, value, unit, tags))

    def get_metrics(
        self,
        name: Optional[str] = None,
        tenant_id: Optional[str] = None,
        since: Optional[datetime] = None,
        limit: int = 1000
    ) -> List[Dict[str, Any]]:
        """Get metrics with optional filtering"""
        filtered_metrics = list(self.metrics)

        if name:
            filtered_metrics = [m for m in filtered_metrics if m.name == name]

        if tenant_id:
            filtered_metrics = [m for m in filtered_metrics if m.tags.get("tenant_id") == tenant_id]

        if since:
            since_str = since.isoformat()
            filtered_metrics = [m for m in filtered_metrics if m.timestamp >= since_str]

        # Convert to dict and limit results
        result = [asdict(m) for m in filtered_metrics[-limit:]]
        return result

    def get_aggregated_metrics(
        self,
        name: str,
        aggregation: str,  # 'sum', 'avg', 'min', 'max', 'count'
        group_by: Optional[str] = None,
        since: Optional[datetime] = None
    ) -> Dict[str, float]:
        """Get aggregated metrics"""
        metrics = self.get_metrics(name, since=since)
        
        if not group_by:
            values = [m["value"] for m in metrics]
            if not values:
                return {}
                
            if aggregation == "sum":
                return {"total": sum(values)}
            elif aggregation == "avg":
                return {"average": sum(values) / len(values)}
            elif aggregation == "min":
                return {"minimum": min(values)}
            elif aggregation == "max":
                return {"maximum": max(values)}
            elif aggregation == "count":
                return {"count": len(values)}
            else:
                return {}

        # Group by specified tag
        grouped = defaultdict(list)
        for metric in metrics:
            key = metric["tags"].get(group_by, "unknown")
            grouped[key].append(metric["value"])

        result = {}
        for key, values in grouped.items():
            if aggregation == "sum":
                result[key] = sum(values)
            elif aggregation == "avg":
                result[key] = sum(values) / len(values)
            elif aggregation == "min":
                result[key] = min(values)
            elif aggregation == "max":
                result[key] = max(values)
            elif aggregation == "count":
                result[key] = len(values)

        return result

    def get_health_metrics(self) -> Dict[str, Any]:
        """Get health check metrics"""
        now = datetime.utcnow()
        five_minutes_ago = datetime.fromtimestamp(now.timestamp() - 5 * 60)
        
        recent_metrics = self.get_metrics(since=five_minutes_ago)
        error_metrics = [m for m in recent_metrics 
                        if "error" in m["name"] or m["tags"].get("status") == "error"]
        
        error_rate = (len(error_metrics) / len(recent_metrics) * 100) if recent_metrics else 0

        checks = {
            "metrics_collecting": len(recent_metrics) > 0,
            "low_error_rate": error_rate < 5,
            "system_responsive": True  # Could add more sophisticated checks
        }

        all_healthy = all(checks.values())
        status = "healthy" if all_healthy else ("unhealthy" if error_rate > 20 else "degraded")

        return {
            "status": status,
            "checks": checks,
            "metrics": {
                "error_rate": error_rate,
                "total_metrics": len(self.metrics),
                "recent_metrics": len(recent_metrics)
            }
        }

# Global metrics collector instance
metrics_collector = MetricsCollector()

File: src/lib/metrics.ts
import { logger } from './logger';

export interface MetricData {
  name: string;
  value: number;
  unit: string;
  timestamp: string;
  tags: Record<string, string>;
}

export interface WorkflowMetrics {
  workflowId: string;
  tenantId: string;
  status: 'started' | 'completed' | 'failed' | 'paused';
  duration?: number;
  stepCount?: number;
  errorCount?: number;
  retryCount?: number;
}

export interface AgentMetrics {
  agentId: string;
  agentType: string;
  tenantId: string;
  workflowId?: string;
  operation: string;
  duration: number;
  status: 'success' | 'error' | 'timeout';
  inputTokens?: number;
  outputTokens?: number;
  ragQueries?: number;
}

export interface SystemMetrics {
  cpuUsage?: number;
  memoryUsage?: number;
  activeWorkflows: number;
  activeTenants: number;
  errorRate: number;
  responseTime: number;
}

class MetricsCollector {
  private metrics: MetricData[] = [];
  private readonly maxMetrics = 10000; // Keep last 10k metrics in memory

  private createMetric(
    name: string,
    value: number,
    unit: string,
    tags: Record<string, string> = {}
  ): MetricData {
    return {
      name,
      value,
      unit,
      timestamp: new Date().toISOString(),
      tags: {
        ...tags,
        environment: process.env.NODE_ENV || 'development'
      }
    };
  }

  private addMetric(metric: MetricData): void {
    this.metrics.push(metric);
    
    // Keep only the most recent metrics
    if (this.metrics.length > this.maxMetrics) {
      this.metrics = this.metrics.slice(-this.maxMetrics);
    }

    // Log metric for external collection
    logger.info('Metric collected', {
      component: 'metrics',
      metadata: metric
    });
  }

  // Workflow metrics
  recordWorkflowStart(workflowId: string, tenantId: string): void {
    this.addMetric(this.createMetric(
      'workflow.started',
      1,
      'count',
      { workflowId, tenantId }
    ));
  }

  recordWorkflowCompletion(metrics: WorkflowMetrics): void {
    const tags = {
      workflowId: metrics.workflowId,
      tenantId: metrics.tenantId,
      status: metrics.status
    };

    this.addMetric(this.createMetric('workflow.completed', 1, 'count', tags));

    if (metrics.duration) {
      this.addMetric(this.createMetric(
        'workflow.duration',
        metrics.duration,
        'milliseconds',
        tags
      ));
    }

    if (metrics.stepCount) {
      this.addMetric(this.createMetric(
        'workflow.steps',
        metrics.stepCount,
        'count',
        tags
      ));
    }

    if (metrics.errorCount) {
      this.addMetric(this.createMetric(
        'workflow.errors',
        metrics.errorCount,
        'count',
        tags
      ));
    }

    if (metrics.retryCount) {
      this.addMetric(this.createMetric(
        'workflow.retries',
        metrics.retryCount,
        'count',
        tags
      ));
    }
  }

  // Agent metrics
  recordAgentOperation(metrics: AgentMetrics): void {
    const tags = {
      agentId: metrics.agentId,
      agentType: metrics.agentType,
      tenantId: metrics.tenantId,
      operation: metrics.operation,
      status: metrics.status
    };

    if (metrics.workflowId) {
      (tags as any).workflowId = metrics.workflowId;
    }

    this.addMetric(this.createMetric(
      'agent.operation',
      1,
      'count',
      tags
    ));

    this.addMetric(this.createMetric(
      'agent.duration',
      metrics.duration,
      'milliseconds',
      tags
    ));

    if (metrics.inputTokens) {
      this.addMetric(this.createMetric(
        'agent.tokens.input',
        metrics.inputTokens,
        'count',
        tags
      ));
    }

    if (metrics.outputTokens) {
      this.addMetric(this.createMetric(
        'agent.tokens.output',
        metrics.outputTokens,
        'count',
        tags
      ));
    }

    if (metrics.ragQueries) {
      this.addMetric(this.createMetric(
        'agent.rag.queries',
        metrics.ragQueries,
        'count',
        tags
      ));
    }
  }

  // System metrics
  recordSystemMetrics(metrics: SystemMetrics): void {
    const tags = { component: 'system' };

    if (metrics.cpuUsage !== undefined) {
      this.addMetric(this.createMetric(
        'system.cpu.usage',
        metrics.cpuUsage,
        'percent',
        tags
      ));
    }

    if (metrics.memoryUsage !== undefined) {
      this.addMetric(this.createMetric(
        'system.memory.usage',
        metrics.memoryUsage,
        'bytes',
        tags
      ));
    }

    this.addMetric(this.createMetric(
      'system.workflows.active',
      metrics.activeWorkflows,
      'count',
      tags
    ));

    this.addMetric(this.createMetric(
      'system.tenants.active',
      metrics.activeTenants,
      'count',
      tags
    ));

    this.addMetric(this.createMetric(
      'system.error.rate',
      metrics.errorRate,
      'percent',
      tags
    ));

    this.addMetric(this.createMetric(
      'system.response.time',
      metrics.responseTime,
      'milliseconds',
      tags
    ));
  }

  // Custom metrics
  recordCustomMetric(
    name: string,
    value: number,
    unit: string,
    tags: Record<string, string> = {}
  ): void {
    this.addMetric(this.createMetric(name, value, unit, tags));
  }

  // Query methods for dashboards
  getMetrics(
    name?: string,
    tenantId?: string,
    since?: Date,
    limit: number = 1000
  ): MetricData[] {
    let filtered = this.metrics;

    if (name) {
      filtered = filtered.filter(m => m.name === name);
    }

    if (tenantId) {
      filtered = filtered.filter(m => m.tags.tenantId === tenantId);
    }

    if (since) {
      filtered = filtered.filter(m => new Date(m.timestamp) >= since);
    }

    return filtered.slice(-limit);
  }

  getAggregatedMetrics(
    name: string,
    aggregation: 'sum' | 'avg' | 'min' | 'max' | 'count',
    groupBy?: string,
    since?: Date
  ): Record<string, number> {
    const metrics = this.getMetrics(name, undefined, since);
    
    if (!groupBy) {
      const values = metrics.map(m => m.value);
      switch (aggregation) {
        case 'sum':
          return { total: values.reduce((a, b) => a + b, 0) };
        case 'avg':
          return { average: values.reduce((a, b) => a + b, 0) / values.length };
        case 'min':
          return { minimum: Math.min(...values) };
        case 'max':
          return { maximum: Math.max(...values) };
        case 'count':
          return { count: values.length };
        default:
          return {};
      }
    }

    const grouped = metrics.reduce((acc, metric) => {
      const key = metric.tags[groupBy] || 'unknown';
      if (!acc[key]) acc[key] = [];
      acc[key].push(metric.value);
      return acc;
    }, {} as Record<string, number[]>);

    const result: Record<string, number> = {};
    for (const [key, values] of Object.entries(grouped)) {
      switch (aggregation) {
        case 'sum':
          result[key] = values.reduce((a, b) => a + b, 0);
          break;
        case 'avg':
          result[key] = values.reduce((a, b) => a + b, 0) / values.length;
          break;
        case 'min':
          result[key] = Math.min(...values);
          break;
        case 'max':
          result[key] = Math.max(...values);
          break;
        case 'count':
          result[key] = values.length;
          break;
      }
    }

    return result;
  }

  // Health check metrics
  getHealthMetrics(): {
    status: 'healthy' | 'degraded' | 'unhealthy';
    checks: Record<string, boolean>;
    metrics: Record<string, number>;
  } {
    const now = new Date();
    const fiveMinutesAgo = new Date(now.getTime() - 5 * 60 * 1000);
    
    const recentMetrics = this.getMetrics(undefined, undefined, fiveMinutesAgo);
    const errorMetrics = recentMetrics.filter(m => 
      m.name.includes('error') || m.tags.status === 'error'
    );
    
    const errorRate = recentMetrics.length > 0 
      ? (errorMetrics.length / recentMetrics.length) * 100 
      : 0;

    const checks = {
      metricsCollecting: recentMetrics.length > 0,
      lowErrorRate: errorRate < 5,
      systemResponsive: true // Could add more sophisticated checks
    };

    const allHealthy = Object.values(checks).every(check => check);
    const status = allHealthy ? 'healthy' : errorRate > 20 ? 'unhealthy' : 'degraded';

    return {
      status,
      checks,
      metrics: {
        errorRate,
        totalMetrics: this.metrics.length,
        recentMetrics: recentMetrics.length
      }
    };
  }
}

export const metricsCollector = new MetricsCollector();
export default metricsCollector;